packages/block-editor/src/hooks/behaviors.js

```diff
@@ -0,0 +1,104 @@
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { SelectControl } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { select } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { InspectorControls } from '../components';
+import { store as blockEditorStore } from '../store';
+
+/**
+ * External dependencies
+ */
+import merge from 'deepmerge';
+
+/**
+ * Override the default edit UI to include a new block inspector control for
+ * assigning behaviors to blocks if behaviors are enabled in the theme.json.
+ *
+ * Currently, only the `core/image` block is supported.
+ *
+ * @param {WPComponent} BlockEdit Original component.
+ *
+ * @return {WPComponent} Wrapped component.
+ */
+export const withBehaviors = createHigherOrderComponent( ( BlockEdit ) => {
+	return ( props ) => {
+		// Only add behaviors to the core/image block.
+		if ( props.name !== 'core/image' ) {
+			return <BlockEdit { ...props } />;
+		}
+
+		const settings =
+			select( blockEditorStore ).getSettings()?.__experimentalFeatures
+				?.blocks?.[ props.name ]?.behaviors;
+
+		if (
+			! settings ||
+			// If every behavior is disabled, do not show the behaviors inspector control.
+			Object.entries( settings ).every( ( [ , value ] ) => ! value )
+		) {
+			return <BlockEdit { ...props } />;
+		}
```

Comment:
There's a subtile issue that may or may not result in a bug. The fact that we conditionally render `BlockEdit` can result in a "remount" of the whole component if the condition above changes from "true" to "false" or the opposite. 

The remount often create focus loss and issues like that. While this might not always be a problem, it is a problem in general if the condition "can" change when the component is mounted (I'm not sure it's the case here)

Reply:
Yeah, probably fine for now, just something to know as sometimes we used to switch over attribute values (easy to add a check there). Also, maybe at some point in the future, that setting could be editable in the global styles UI in the site editor.

Reply:
> that setting could be editable in the global styles UI in the site editor

That is the next step after merging this PR, so we should then take a look at this issue.